src/cast.rs

```diff
@@ -0,0 +1,958 @@
+
+  fn into_raw(self) -> Self::Raw {
+    self
+  }
+  unsafe fn from_raw(data: Self::Raw) -> Self {
+    &mut *data
+  }
+}
+
+unsafe impl<'a, T: 'a> Container<'a> for &'a [T] {
+  type Class = RefT;
+  type Item = T;
+  type Raw = (*const T, usize);
+  type CastErr = OnlyErr;
+
+  fn into_raw(self) -> Self::Raw {
+    (self.as_ptr(), self.len())
+  }
+  unsafe fn from_raw((data, len): Self::Raw) -> Self {
+    slice::from_raw_parts(data, len)
+  }
+}
+
+unsafe impl<'a, T: 'a> Container<'a> for &'a mut [T] {
+  type Class = RefT;
+  type Item = T;
+  type Raw = (*mut T, usize);
+  type CastErr = OnlyErr;
+
+  fn into_raw(self) -> Self::Raw {
+    (self.as_mut_ptr(), self.len())
+  }
+  unsafe fn from_raw((data, len): Self::Raw) -> Self {
+    slice::from_raw_parts_mut(data, len)
+  }
+}
+
+unsafe impl<'a, T: 'a> Container<'a> for *const T {
+  type Class = PtrT;
+  type Item = T;
+  type Raw = Self;
+  type CastErr = OnlyErr;
+
+  fn into_raw(self) -> Self::Raw {
+    self
+  }
+  unsafe fn from_raw(ptr: Self::Raw) -> Self {
+    ptr
+  }
+}
+
+unsafe impl<'a, T: 'a> Container<'a> for *mut T {
+  type Class = PtrT;
+  type Item = T;
+  type Raw = Self;
+  type CastErr = OnlyErr;
+
+  fn into_raw(self) -> Self::Raw {
+    self
+  }
+  unsafe fn from_raw(ptr: Self::Raw) -> Self {
+    ptr
+  }
+}
+
+unsafe impl<'a, T: 'a> Container<'a> for NonNull<T> {
+  type Class = NonNullT;
+  type Item = T;
+  type Raw = *mut T;
+  type CastErr = OnlyErr;
+
+  fn into_raw(self) -> Self::Raw {
+    self.as_ptr()
+  }
+  unsafe fn from_raw(ptr: Self::Raw) -> Self {
+    NonNull::new_unchecked(ptr)
+  }
+}
+
+#[cfg(feature = "non_null_slice_cast")]
+unsafe impl<'a, T: 'a> Container<'a> for NonNull<[T]> {
+  type Class = NonNullT;
+  type Item = T;
+  type Raw = (*mut T, usize);
+  type CastErr = OnlyErr;
+
+  fn into_raw(self) -> Self::Raw {
+    (self.as_ptr() as *mut T, self.len())
+  }
+  unsafe fn from_raw((data, len): Self::Raw) -> Self {
+    NonNull::new_unchecked(core::ptr::slice_from_raw_parts_mut(data, len))
+  }
+}
+
+unsafe impl<'a, T: 'a> Container<'a> for AtomicPtr<T> {
+  type Class = AtomicPtrT;
+  type Item = T;
+  type Raw = *mut T;
+  type CastErr = OnlyErr;
+
+  fn into_raw(self) -> Self::Raw {
+    self.into_inner()
+  }
+  unsafe fn from_raw(ptr: Self::Raw) -> Self {
+    Self::new(ptr)
+  }
+}
+
+unsafe impl<'a, C> Container<'a> for Option<C>
+where
+  C: Container<'a>,
+  C::CastErr: CastErrWithValue<Self>,
+{
+  type Class = OptionT<C::Class>;
+  type Item = C::Item;
+  type Raw = Option<C::Raw>;
+  type CastErr = C::CastErr;
+
+  fn into_raw(self) -> Self::Raw {
+    self.map(|x| x.into_raw())
+  }
+  unsafe fn from_raw(raw: Self::Raw) -> Self {
+    raw.map(|raw| C::from_raw(raw))
+  }
+}
+
+// SAFETY: `Pin` has no safety requirements for types which deref to an `Unpin` type.
+unsafe impl<'a, C> Container<'a> for Pin<C>
+where
+  C: Container<'a> + Deref<Target = <C as Container<'a>>::Item>,
+  C::Item: Unpin,
```

Comment:
Requiring `C::Target = C::Item` prevents casting things like `Pin<Box<[T]>>`, since `<Box<[T]> as Container>::Item = T`, but `<Box<[T]> as Deref>::Target = [T]`
```suggestion
  C: Container<'a> + Deref,
  C::Target: Unpin,
  C::Item: Unpin,
```

Reply:
The change should go the other way with `Box<[T]>` setting the correct item type.